Added #MajorItem item group for plando

cjohnson57 opened the pull request:
Item group is settings dependent, so if Gerudo card is not shuffled or GF is open it will not place Gerudo card, it won't place Triforce pieces unless Triforce Hunt is on, etc.

cjohnson57 committed:
Added #MajorItem item group for plando

cjohnson57 committed:
Added checks for settings I forgot about; reordered settings checks

cjohnson57 committed:
Fixed bugs with weird eggs in item pool

flagrama commented on Plandomizer.py:
```diff
@@ -379,30 +379,38 @@ def alter_pool(self, world, pool):
             if record.type == 'remove':
                 self.pool_remove_item([pool], item_name, record.count)
 
+        remove_egg = False
```
1, I'd not bother with this

flagrama commented on Plandomizer.py:
```diff
@@ -379,30 +379,38 @@ def alter_pool(self, world, pool):
+                                bottles -= 1
+                            elif trade_matcher(item):
+                                self.pool_add_item(pool, "#AdultTrade", 1)           
+        if remove_egg:
+            del self.item_pool['Weird Egg']
```
2. and move this

cjohnson57 replied:
So for 1. and 2., it had to be done that way because you can't remove an element of a dict while iterating through that dict

flagrama replied:
Ah, that makes sense.

flagrama commented on Plandomizer.py:
```diff
@@ -379,30 +379,38 @@ def alter_pool(self, world, pool):
-                        elif trade_matcher(item):
-                            self.pool_remove_item([pool], "#AdultTrade", 1)
+                elif item_name == 'Weird Egg' and not self.settings.shuffle_weird_egg:
+                    remove_egg = True
```
3. to here, and in both of the instances of it being Weird Egg, `continue` the loop instead of making the rest of the loop an else statement.

cjohnson57 replied:
That is much cleaner, I didn't know that was a thing in Python, will do

cjohnson57 replied:
ok done!

cjohnson57 committed:
Cleaned up loop with continues as suggested by flagrama

Roman971 commented:
I assume the intention is to have a group that corresponds to all items considered "major" in a given set of settings, so you should be able to use the `majoritem` property from the `Item` class to avoid hardcoding/duplicating every condition that makes an item major or not.

cjohnson57 commented:
> I assume the intention is to have a group that corresponds to all items considered "major" in a given set of settings, so you should be able to use the `majoritem` property from the `Item` class to avoid hardcoding/duplicating every condition that makes an item major or not.

I have 2 issues with this approach:
1) I'm not sure if that tag can be accessed from the lists where the item groups are set in ItemPool.py, I think that code is run before settings are even checked but not 100% sure on that.
2) An item can be major either way, but that alone does not determine if it should be placed by the item group, it's more a matter of if it's shuffled or not. For example gerudo card is always major as long as GF isn't open, but if the card isn't shuffled and thus set to its default location, you don't want to also place it elsewhere in the world.

cjohnson57 committed:
Added unit test for weird egg fix

Zannick reviewed (changes requested):
> I'm not sure if that tag can be accessed from the lists where the item groups are set in ItemPool.py, I think that code is run before settings are even checked but not 100% sure on that.

I think the issue is actually that this function expects item names (strings) rather than objects.

> An item can be major either way, but that alone does not determine if it should be placed by the item group, it's more a matter of if it's shuffled or not. For example gerudo card is always major as long as GF isn't open, but if the card isn't shuffled and thus set to its default location, you don't want to also place it elsewhere in the world.

This should be determined by the item pool, not the group definition. (See for example, Gerudo card can be placed a second time in plentiful.) 
https://github.com/TestRunnerSRL/OoT-Randomizer/blob/02be6c494c8ad2b3acd5f35c9a5ebed1bb4fb0e3/ItemPool.py#L1086-L1087
Does removing the extra ifs in `pattern_matcher` cause items to be added when they shouldn't be? (Note this would obviate the comments I put on the code, if we can just remove the use of settings in `pattern_matcher`.)

cjohnson57 committed:
Merge remote-tracking branch 'upstream/Dev' into Dev

Zannick commented on Plandomizer.py:
```diff
@@ -1063,18 +1073,37 @@ def is_pattern(pattern):
 
 
-def pattern_matcher(pattern):
+def pattern_matcher(self, pattern):
```
Here's what I'd propose instead: generate this group list in WorldDistribution on init, move pattern_matcher into WorldDistribution, and then if pattern is `#MajorItem` we have the group already built. (This should save list modification on every call to `pattern_matcher`.)

Zannick replied:
Of course, as discussed in Discord, the other alternative is to change the predicate passed to ItemIterator in `pool_add_item` and `pool_replace_item` to check `item.majoritem` (and potentially another property specifically to avoid placing items like GF when it's in a fixed place, since `.majoritem` doesn't cover all of this).

cjohnson57 committed:
Modified major item settings matching to use base pool

Also changed to only pass base_pool through to pattern_matcher rather than the entire Distribution object.

cjohnson57 committed:
Made functions members of WorldDistribution to make everything cleaner

Also added major_group that's only computed once (if necessary at all)

cjohnson57 committed:
Added songs and trade items to group

cjohnson57 committed:
Added special handling for triforce pieces, keys, and tokens
